### DISCUSSION TITLE
03 Pytorch Computer Vision

### DISCUSSION THREAD

**Discussion by ZaidKhan-git** [mrdbourke/pytorch-deep-learning#1376] [Q&A]
I am confused how does ```next(iter(train_dataloader))``` returns train_features_batch and train_labels_batch ? 
Also if dataloader already returns an iterable, then why are we making it iterable again ? I am bit confused here.

**Comment by tt-a1i** [comment]
`DataLoader` is iterable, but it isn't the iterator itself.

`iter(train_dataloader)` creates an iterator, then `next(...)` asks that iterator for one batch. Since each dataset item contains a feature and a label, the batch also contains two values, which can be unpacked like this:

```python
train_features_batch, train_labels_batch = next(iter(train_dataloader))
```

A `for` loop does the same work automatically:

```python
for train_features_batch, train_labels_batch in train_dataloader:
    break
```

You only need to call `iter()` yourself when you want to manually fetch a batch with `next()`.
